src/entity.zig

```diff
@@ -225,34 +227,38 @@ pub const ClientEntityManager = struct {
 			}
 		}
 	}
-
 	pub fn serverUpdate(time: i16, data: []const u8) void {
+		_serverUpdate(time, data) catch |err| {
+			std.log.err("Corrupted entity update packet ({s})", .{@errorName(err)});
+		};
+	}
+
+	fn _serverUpdate(time: i16, data: []const u8) !void {
 		mutex.lock();
 		defer mutex.unlock();
+		if(data.len%(4 + 24 + 12 + 24) != 0) return error.corrupted;
+
 		timeDifference.addDataPoint(time);
-		std.debug.assert(data.len%(4 + 24 + 12 + 24) == 0);
-		var remaining = data;
-		while(remaining.len != 0) {
-			const id = std.mem.readInt(u32, remaining[0..4], .big);
-			remaining = remaining[4..];
+		var reader = BinaryReader.init(data, .big);
+
+		while(reader.remaining.len != 0) {
+			const id = try reader.readInt(u32);
 			const pos = [_]f64{
-				@bitCast(std.mem.readInt(u64, remaining[0..8], .big)),
-				@bitCast(std.mem.readInt(u64, remaining[8..16], .big)),
-				@bitCast(std.mem.readInt(u64, remaining[16..24], .big)),
-				@floatCast(@as(f32, @bitCast(std.mem.readInt(u32, remaining[24..28], .big)))),
-				@floatCast(@as(f32, @bitCast(std.mem.readInt(u32, remaining[28..32], .big)))),
-				@floatCast(@as(f32, @bitCast(std.mem.readInt(u32, remaining[32..36], .big)))),
+				try reader.readFloat(f64),
```

Comment:
It was the same series of casts copied 9 times, IMO good place to introduce new `readFloat` function.

Reply:
yes, and please also apply these changes (in a separate PR) to all the other code that used the float bitcast.